crates/red_knot_python_semantic/src/types/unpacker.rs

```diff
@@ -1,126 +1,185 @@
 use std::borrow::Cow;
 
+use itertools::Either;
+use rustc_hash::FxHashMap;
+
 use ruff_db::files::File;
 use ruff_python_ast::{self as ast, AnyNodeRef};
-use rustc_hash::FxHashMap;
 
 use crate::semantic_index::ast_ids::{HasScopedExpressionId, ScopedExpressionId};
 use crate::semantic_index::symbol::ScopeId;
 use crate::types::{todo_type, Type, TypeCheckDiagnostics};
 use crate::Db;
 
 use super::context::{InferContext, WithDiagnostics};
+use super::{TupleType, UnionType};
 
 /// Unpacks the value expression type to their respective targets.
 pub(crate) struct Unpacker<'db> {
     context: InferContext<'db>,
+    scope: ScopeId<'db>,
     targets: FxHashMap<ScopedExpressionId, Type<'db>>,
 }
 
 impl<'db> Unpacker<'db> {
-    pub(crate) fn new(db: &'db dyn Db, file: File) -> Self {
+    pub(crate) fn new(db: &'db dyn Db, file: File, scope: ScopeId<'db>) -> Self {
         Self {
             context: InferContext::new(db, file),
             targets: FxHashMap::default(),
+            scope,
         }
     }
 
     fn db(&self) -> &'db dyn Db {
         self.context.db()
     }
 
-    pub(crate) fn unpack(&mut self, target: &ast::Expr, value_ty: Type<'db>, scope: ScopeId<'db>) {
+    /// Unpack the value type to the target expression.
+    pub(crate) fn unpack(&mut self, target: &ast::Expr, value_ty: Type<'db>) {
+        debug_assert!(
+            matches!(target, ast::Expr::List(_) | ast::Expr::Tuple(_)),
+            "Unpacking target must be a list or tuple expression"
+        );
+
+        self.unpack_inner(target, value_ty);
+    }
+
+    fn unpack_inner(&mut self, target: &ast::Expr, value_ty: Type<'db>) {
```

Comment:
Refactor 1: Add a debug assertion to make sure that the target is a list or a tuple.